tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+team. Do not provide the design details in this, section - there is a
+dedicated section for that later in the document.
+-->
+
+The Radius Flux Controller is a new Radius control plane service that, along with the Radius DeploymentTemplate Controller, enables GitOps practices for Radius users. It is responsible for reading the Git repository, compiling Bicep files, and creating the DeploymentTemplate resources on the cluster. This controller ensures that the infrastructure defined in the Git repository is accurately reflected in the Kubernetes cluster, enabling a seamless GitOps workflow.
+
+## Terms and definitions
+
+<!--
+Include any terms, definitions, or acronyms that are used in
+this design document to assist the reader. They may or may not
+be part of the user-facing experience once implemented, and can
+be specific to this design context.
+-->
+
+- **GitOps**: A set of practices to manage infrastructure and application configurations using Git.
+- **Bicep**: A domain-specific language (DSL) for deploying Azure resources declaratively.
+- **DeploymentTemplate**: A custom resource in Kubernetes that represents the desired state of a Bicep deployment.
+
+## Objectives
+
+<!--
+Describe goals/non-goals and user-scenario of this feature to understand
+the end-user goals.
+* If the feature shares the same objectives of the existing design, link
+  to the existing doc and section rather than repeat the same context.
+* If the feature has a scenario, UX, or other product feature design doc,
+  link it here and summarize the important parts.
+-->
+
+> **Issue Reference:** #6689
+
+### Goals
+
+<!--
+Describe goals to define why we are doing this work, how we will make
+priority decisions, and how we will determine success.
+-->
+
+- Enable seamless integration of GitOps workflows with Radius + Bicep. Users should be able to specify their Bicep files and parameters in a Git repository and have the resources within them automatically deployed to their Kubernetes cluster.
+- Ensure that the infrastructure defined in the Git repository is accurately reflected in the Kubernetes cluster. This includes creating, updating, and deleting resources as necessary, based on the changes in the Git repository.
+- Provide an experience that is consistent with existing Flux workflows. Users should be able to use Flux as they do today, with the addition of the Radius Flux Controller for managing Bicep files.
+
+### Non goals
+
+<!--
+Describe non-goals to identify something that we won’t be focusing on
+immediately. We won’t be expending any effort on these matters. If there
+will be follow-ups after this work, list them here. If there are things
+we plan to do in the future, but are out of scope of this design, list
+them here. Provide a brief explanation on why this is a non-goal.
+-->
+
+- Supporting non-Flux GitOps tools (e.g., ArgoCD).
+- Supporting non-Bicep IaC formats.
+- Radius Dashboard support
```

Comment:
Can you elaborate on what specifically won't be supported in Radius Dashboard? I'm assuming gitops "clickops" via dashboard UI won't be supported, but things like App Graph will still get updated when resources are changed by Flux-triggered operations?

Reply:
I'll elaborate here. I meant that "Radius GitOps" (status of DeploymentTemplate resources, GitRepositories, etc.) will not be tracked in the dashboard as part of this work. Everything Radius users have today will still be supported.